File: final/src/board.rs
use std::vec::Vec;

pub struct Board {
    pub size: usize,
    pub nodes: Vec<Node>,
}

impl Board {
    pub fn new(size: usize, board: [[u8; 4]; 4]) -> Board {
        let mut nodes: Vec<Node> = Vec::with_capacity(size * size);

        for i in 0..size * size {
            let new_node: Node = Node::new(size, board, i);

            nodes.push(new_node);
        }

        return Board {
            size: size,
            nodes: nodes,
        };
    }

    pub fn print_board(&self) {
        for i in 0..self.nodes.len() {
            let node: &Node;

            match self.nodes.get(i) {
                Some(n) => node = n,
                None => panic!("err in print_board"),
            }

            if node.index % self.size == 0 {
                println!();
            }

            print!("{num:>2} ", num = node.value);
        }
    }

    pub fn update_possibles(&self) {
        for node in self.nodes.iter() {
            print!("{}-", node.value);

            for i in node.sqr.iter() {
                print!("{} ", i);
            }

            // if node.solved {
            // }

            println!();
        }

        // for i in 0..self.nodes.len() {
        //     let node: &Node = self.get_node(i);

        //     if node.solved {
        //         for i in 0..self.size {
        //             let node_u: &Node;
        //             let node_col_idx: usize;
        //             // let node_row: &Node;
        //             // let node_sqr: &Node;

        //             match node.col.get(i) {
        //                 Some(n) => node_col_idx = *n,
        //                 None => panic!("err in update_possibilities"),
        //             }

        //             match self.nodes.get_mut(node_col_idx) {
        //                 Some(n) => node_u = n,
        //                 None => panic!("FFS"),
        //             }

        //             // match self.nodes.get_mut(i) {
        //             //     Some(n) => Node::remove_possibility(n, node.value),
        //             //     None => println!("Err"),
        //             // }
        //         }
        //     }
        // }
    }

    fn get_node(&self, index: usize) -> &Node {
        match self.nodes.get(index) {
            Some(n) => return n,
            None => panic!("err in update_possibilities"),
        }
    }
}

pub struct Node {
    pub possibles: Vec<u8>,
    pub value: u8,
    pub row: Vec<usize>,
    pub col: Vec<usize>,
    pub sqr: Vec<usize>,
    pub solved: bool,
    index: usize,
}

impl Node {
    pub fn new(size: usize, board: [[u8; 4]; 4], index: usize) -> Node {
        let mut possibles: Vec<u8> = Vec::with_capacity(size);
        let mut row: Vec<usize> = Vec::with_capacity(size);
        let mut col: Vec<usize> = Vec::with_capacity(size);
        let mut sqr: Vec<usize> = Vec::with_capacity(size);
        let x_y = index_to_x_y(index, size);
        let x = x_y.0;
        let y = x_y.1;
        let value = board[x as usize][y as usize];
        let mut solved: bool = false;

        // println!("{} -> ({},{})", index, x, y);

        if value != 0 {
            possibles.push(value);
            solved = true;
        }

        for i in 0..size {
            if value == 0 {
                possibles.push((i + 1) as u8);
            } //Pushed possible values

            if i as u8 != x {
                row.push(x_y_to_index(x, i as u8, size));
            } //Pushes everything in the row
            if i as u8 != y {
                col.push(x_y_to_index(i as u8, y, size));
            } //Pushes everyting in the col
        }

        //Square
        let sub_size = (size as f64).sqrt() as u8;
        let sub_x = x_y.0 / sub_size;
        let sub_y = x_y.1 / sub_size;

        for x_i in 0..sub_size {
            for y_i in 0..sub_size {
                let n_index = x_y_to_index(sub_size * sub_x + x_i, sub_size * sub_y + y_i, size);

                if index != n_index {
                    sqr.push(n_index);
                }
            }
        }

        return Node {
            possibles: possibles,
            value: value,
            row: row,
            col: col,
            sqr: sqr,
            solved: solved,
            index: index,
        };
    }

    fn remove_possibility(&mut self, val: u8) {
        match self.possibles.iter().position(|&x| x == val) {
            Some(x) => self.possibles.swap_remove(x),
            None => return,
        };
    }

    pub fn print_val(&mut self) {
        print!("{}", self.value);
    }
}

fn x_y_to_index(x: u8, y: u8, size: usize) -> usize {
    return ((y as usize * size) + x as usize) as usize;
}

fn index_to_x_y(index: usize, size: usize) -> (u8, u8) {
    let row = index / size;
    return ((index - (row * size)) as u8, row as u8);
}

File: final/js/board.js
let NodeImp = require("./node.js");
let { Node } = NodeImp;

class Board {
    /*
        Constructor
    */
    constructor(board) {
        //console.log(board);
        this.nodes = [];
        this.size = board.length;

        for (var i = 0; i < this.size * this.size; i++) {
            let { x, y } = index_to_x_y(i, this.size);
            let new_node = new Node(this.size, x, y, board[y][x], i);
            this.nodes.push(new_node);
        }

        //Updates all nodes to include changes for initial values
        this.nodes.forEach(node => {
            if (node.solved) {
                const value = node.value;

                node.row.forEach(index => {
                    if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                        this.propagate_update(index);
                    }
                });

                node.col.forEach(index => {
                    if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                        this.propagate_update(index);
                    }
                });

                node.sqr.forEach(index => {
                    if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                        this.propagate_update(index);
                    }
                });
            }
        });

        //Looks for sole candidates
        while (true) {
            let count = this.examine_possibles(false);
            //console.log(`Updated ${count}`);
            if (count === 0) break;
        };
    }

    /*
        Propagate Update

        When a node is updated (to solved) then this will propagate the new constraints
     */
    propagate_update(index) {
        let nodes = this.nodes;
        const node = this.nodes[index];
        const value = node.value;

        if (!node.solved) return;

        node.row.forEach(index => {
            if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                this.propagate_update(index);
            }
        });
        node.col.forEach(index => {
            if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                this.propagate_update(index);
            }
        });
        node.sqr.forEach(index => {
            if (this.nodes[index].remove_possibility(value) && this.is_valid_solve(this.nodes[index])) {
                this.propagate_update(index);
            }
        });


    }

    /*
        Is Valid Solve

        Checks if a node is actually being solved properly
     */
    is_valid_solve(node) {
        const value = node.possibles[0];

        for (var i = 0; i < node.row.length; i++) {
            if (this.nodes[node.row[i]].value == value) return false;
            if (this.nodes[node.col[i]].value == value) return false;
            if (this.nodes[node.sqr[i]].value == value) return false;
        }

        return true;
    }

    /*
        Examine Possibles

        Looks if a cell is the only candidate for a number in one of its subdomains

        if parameter count is true, no updates will occur will only count number of changes (for heuristic)
    */
    examine_possibles(count) {
        let updates = 0;
        //Loop through diagonally (hitting all rows/columns in n iterations)
        for (var i = 0; i < this.size; i++) {
            let node = this.nodes[i * (this.size + 1)];
            updates += this.calculate_frequencies(node, 0, count); //Row
            updates += this.calculate_frequencies(node, 1, count); //Col
        }

        let inc = Math.sqrt(this.size);

        //Loop through top left of all sub_squares
        for (var y = 0; y < this.size; y += inc) {
            for (var x = 0; x < this.size; x += inc) {
                updates += this.calculate_frequencies(this.nodes[x_y_to_index(x, y, this.size)], 2, count); //Sqr
            }
        }

        return updates;
    }

    /*
        ECalculate Frequencies

        Examine possibles helper (does the sole candidate on a per node basis);
    */
    calculate_frequencies(node, choice, count) {
        let updates = 0;
        let freq = [];
        let arr = (choice === 0) ? node.row : (choice === 1) ? node.col : node.sqr;

        if (node.solved) {
            freq[node.value] = { solved: true };
        }
        else {
            add_all_possibles(node);
        }

        arr.forEach(index => {
            let sub_node = this.nodes[index];

            if (sub_node.solved) {
                freq[sub_node.value] = { solved: true };
                return;
            }

            add_all_possibles(sub_node);
        });

        for (var i = 1; i < freq.length; i++) {
            if (freq[i] == undefined) continue;
            if (!freq[i].solved && freq[i].indices.length === 1) {
                //console.log(`Index:${freq[i].indices[0]} is now ${i}`);
                updates++;
                if (!count) {
                    this.nodes[freq[i].indices[0]].solve(i);
                    this.propagate_update(freq[i].indices[0]);
                }
            }
        }

        return updates;

        function add_all_possibles(node) {
            node.possibles.forEach(pos => {
                if (freq[pos] == undefined) {
                    freq[pos] = {
                        indices: [node.index],
                        solved: false
                    }
                }
                else if (!freq[pos].solved) {
                    freq[pos].indices.push(node.index);
                }
            });
        }
    }

    /*
        Is Valid

        Checks if the board is valid
    */
    is_valid(debug) {
        //Loop through diagonally (hitting all rows/columns in n iterations)
        for (var i = 0; i < this.size; i++) {
            let node = this.nodes[i * (this.size + 1)];
            let row_seen = {};
            let col_seen = {};

            if (node.solved) {
                row_seen[node.value] = { i: node.index };
                col_seen[node.value] = { i: node.index };
            }

            for (var i = 0; i < node.row.length; i++) {
                const node_row = this.nodes[node.row[i]];
                const node_col = this.nodes[node.col[i]];

                if (node_row.solved) {
                    if (row_seen[node_row.value]) {
                        if (debug) {
                            let n = this.nodes[row_seen[node_row.value].i];
                            console.log(`Row - Value: ${n.value} - Node: ${n.index}=(${index_to_x_y(n.index, this.size).x},${index_to_x_y(n.index, this.size).y})`)
                            console.log(`Row - Value: ${node_row.value} - Node: ${node_row.index}=(${index_to_x_y(node_row.index, this.size).x},${index_to_x_y(node_row.index, this.size).y})`)
                        }
                        return false;
                    }
                    row_seen[node_row.value] = { i: node_row.index };
                }

                if (node_col.solved) {
                    if (col_seen[node_col.value]) {
                        if (debug) {
                            let n = this.nodes[col_seen[node_col.value].i];
                            console.log(`Col - Value: ${n.value} - Node: ${n.index}=(${index_to_x_y(n.index, this.size).x},${index_to_x_y(n.index, this.size).y})`)
                            console.log(`Col - Value: ${node_col.value} - Node: ${node_col.index}=(${index_to_x_y(node_col.index, this.size).x},${index_to_x_y(node_col.index, this.size).y})`)
                        }
                        return false;
                    }

                    col_seen[node_col.value] = { i: node_col.index };
                }
            }
        }

        let inc = Math.sqrt(this.size);

        //Loop through top left of all sub_squares
        for (var y = 0; y < this.size; y += inc) {
            for (var x = 0; x < this.size; x += inc) {
                let sqr_seen = {};
                let node = this.nodes[x_y_to_index(x, y, this.size)];

                if (node.solved) sqr_seen[node.value] = { i: node.index }

                for (var i = 0; i < node.row.length; i++) {
                    const node_sqr = this.nodes[node.sqr[i]];

                    if (node_sqr.solved) {
                        if (sqr_seen[node_sqr.value]) {
                            if (debug) {
                                let n = this.nodes[sqr_seen[node_sqr.value].i];
                                console.log(`Sqr - Value: ${n.value} - Node: ${n.index}=(${index_to_x_y(n.index, this.size).x},${index_to_x_y(n.index, this.size).y})`)
                                console.log(`Sqr - Value: ${node_sqr.value} - Node: ${node_sqr.index}=(${index_to_x_y(node_sqr.index, this.size).x},${index_to_x_y(node_sqr.index, this.size).y})`)
                            }

                            return false;
                        }
                        sqr_seen[node_sqr.value] = { i: node_sqr.index };
                    }
                }
            }
        }

        return true;
    }

    /*
        Is Solved
     
        Checks if the board is solved
    */
    is_solved() {
        for (var i = 0; i < this.nodes.length; i++) {
            if (!this.nodes[i].solved) return false;
        }

        //return this.is_valid();
        return true;
    }

    /*
        List possibilities
     
        Debugging function to list possible answers per node
    */
    list_possibles() {
        this.nodes.forEach(node => {
            if (node.solved) return;
            console.log(`${node.index}: ${node.possibles}`);
        });
    }

    /*
        Print Board
    */
    print_board() {
        let str = "";
        let zeroes = 0;

        this.nodes.forEach(node => {
            if (node.index % this.size == 0) {
                console.log(`${str}`);
                str = "";
            }

            str += " " + node.value;
            if (node.value === 0) zeroes++;
        });

        console.log(`${str}\nZeroes:${zeroes}`);
    }

    /*
        Get Board
     
        Returns the board in a 2d array (like board templates);
    */
    get_board() {
        let board = [];
        let row = [];
        this.nodes.forEach(node => {
            row.push(node.value);

            if ((node.index + 1) % this.size == 0) {
                board.push(row);
                row = [];
            }
        });

        return board;
    }

    get_node(index) {
        if (index < 0 || index >= this.nodes.length) return;

        return this.nodes[index];
    }

    get_neighbours() {
        let neighbours = [];
        let min = { size: Number.MAX_SAFE_INTEGER, index: 0 };

        for (var n = 0; n < this.nodes.length; n++) {
            let node = this.nodes[n];
            if (node.solved) continue;
            if (node.possibles.length < min.size) min = { size: node.possibles.length, index: node.index };
        }

        let node = this.nodes[min.index];

        //console.log(min, node.possibles);

        for (var i = 0; i < min.size; i++) {
            let board = this.get_board();
            if (board[node.y][node.x].solved) console.log("Broken");
            board[node.y][node.x] = node.possibles[i];
            neighbours.push(board);
        }

        return neighbours;
    }

    get_all_min_neighbours() {
        let neighbours = [];
        let min = this.size;

        for (var n = 0; n < this.nodes.length; n++) {
            let node = this.nodes[n];
            if (node.solved) continue;
            if (node.possibles.length < min) min = node.possibles.length;
        }

        for (var n = 0; n < this.nodes.length; n++) {
            let node = this.nodes[n];
            if (!node.solved && node.possibles.length === min)
                for (var i = 0; i < min; i++) {
                    let board = this.get_board();
                    if (board[node.y][node.x].solved) console.log("Broken");
                    board[node.y][node.x] = node.possibles[i];
                    neighbours.push(board);
                }
        }

        return neighbours;
    }
}

function x_y_to_index(x, y, size) {
    return y * size + x;
}

function index_to_x_y(index, size) {
    let row = Math.floor(index / size);
    return {
        x: (index - (row * size)), y: row
    };
}

module.exports = {
    Board
}
